src/access/RequiredPrivilege.cpp

```diff
@@ -0,0 +1,214 @@
+    };
+
+PrivilegeOverride
+    privilegeOverrideForWriteAttribute[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_WRITE_ATTRIBUTE] = {
+        PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+    };
+
+PrivilegeOverride
+    privilegeOverrideForInvokeCommand[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_INVOKE_COMMAND] = {
+        PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+    };
+
+PrivilegeOverride privilegeOverrideForReadEvent[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_READ_EVENT] = {
+    PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+};
+
+enum class Operation
+{
+    kReadAttribute  = 0,
+    kWriteAttribute = 1,
+    kInvokeCommand  = 2,
+    kReadEvent      = 3
+};
+
+PrivilegeOverride * const privilegeOverride[] = { privilegeOverrideForReadAttribute, privilegeOverrideForWriteAttribute,
+                                                  privilegeOverrideForInvokeCommand, privilegeOverrideForReadEvent };
+
+const size_t maxPrivilegeOverride[] = { sizeof(privilegeOverrideForReadAttribute) / sizeof(privilegeOverrideForReadAttribute[0]),
+                                        sizeof(privilegeOverrideForWriteAttribute) / sizeof(privilegeOverrideForWriteAttribute[0]),
+                                        sizeof(privilegeOverrideForInvokeCommand) / sizeof(privilegeOverrideForInvokeCommand[0]),
+                                        sizeof(privilegeOverrideForReadEvent) / sizeof(privilegeOverrideForReadEvent[0]) };
+
+Privilege GetRequiredPrivilege(Operation operation, ClusterId cluster, EndpointId endpoint, FieldId field,
+                               Privilege defaultPrivilege)
+{
+    VerifyOrDie(cluster != kInvalidClusterId && endpoint != kInvalidEndpointId && field != kInvalidFieldId);
+
+    const auto * const pStart = privilegeOverride[static_cast<int>(operation)];
+    const auto * const pEnd   = pStart + maxPrivilegeOverride[static_cast<int>(operation)];
+
+    // TODO: note that the sorted nature of the table can be taken advantage of to skip around
+    // (e.g. binary search), but if so, ensure to look for specific (non-wilcard) entries before
+    // using a wildcard entry.
+
+    for (const auto * p = pStart; p < pEnd; ++p)
+    {
+        if (p->mCluster == kInvalidClusterId || p->mCluster > cluster)
+        {
+            break; // not found
+        }
+        if (p->mCluster == cluster && (p->mEndpoint == endpoint || p->mEndpoint == kInvalidEndpointId) &&
+            (p->mField == field || p->mField == kInvalidFieldId))
+        {
+            return p->mPrivilege;
+        }
+    }
+
+    return defaultPrivilege;
+}
+
+CHIP_ERROR OverrideRequiredPrivilege(Operation operation, ClusterId cluster, EndpointId endpoint, FieldId field,
+                                     Privilege privilege)
```

Comment:
Why do we want this API?  These tables represent static spec-defined data... Is this to handle device type overrides of cluster specs, which we can't generate statically because we don't know what device type will be on which endpoint?

I'd say we should not worry about that until we actually have some device type specs overriding cluster spec privileges.  Which so far we don't, iirc.

Reply:
Well, I figured we'd have a few levels of it.

Some could put their overrides in this table, which ideally is const (and you noticed it isn't now). Some could add to it at startup, e.g. in their cluster init function in the server.cpp. And some would have to manage it dynamically at runtime (I was thinking of bridges).

We can talk about how we want to do this, I just put this up as a straw-man to provoke such discussion, and as a skeleton that others could take to flesh out more.

Reply:
I removed the dynamic APIs, someone can write them if/when we need them.